enhancements/cli-in-cluster-management.md

```diff
@@ -0,0 +1,177 @@
+Some current examples are oc, kubectl, odo, istio, tekton, and knative.  It is difficult for users to discover what tools exist, 
+where to download them from and which version they should download.  We need to simplify as much as possible the interaction
+of services on OpenShift.  We need a mechanism for providing and consuming tools that is simple to add on to as new CLIs are 
+developed from a variety of sources - and this needs to be managed from within a cluster and with disconnected installs.
+This is the motivation for the CLI in-cluster manager.
+
+### Goals
+
+Each component wishing to provide customers with their binaries in-cluster will build and publish artifacts via an official channel to a central index
+from which the CLI in-cluster manager can access them. 
+
+Possible routes for supplying CLI binaries:    
+
+1.  Central index that has links to where artifacts are stored.  This is what we have now with
+[the ConsoleCLIDownload CRD](https://github.com/openshift/api/blob/master/console/v1/0000_10_consoleclidownload.crd.yaml).
+Currently, a user can run `oc get consoleclidownloads oc-cli-downloads -o yaml` or `oc get consoleclidownloads odo-cli-downloads -o yaml` 
+to get a download link for `odo` or `oc`.  The in-cluster cli-manager will take this a step further to handle installing binaries on the user's system, 
+in the $HOME directory, that are known compatible with a cluster's version.  The challenge is that we need a mechanism that works in disconnected environments.
+Also, this has created a burden of maintenance for the console team.  Furthermore, we want to offer a mechanism that other CLI authors can leverage to deliver tools in-cluster.  
+CLI-manager will replace the current consoleclidownloads mechanism and will extend its function.
+
+2.  Central repository where all artifacts are stored - [index of /pub/openshift-v4/clients](https://mirror.openshift.com/pub/openshift-v4/clients/)
+currently is where crc, oc, ocp-dev-preview, ocp, odo, and serverless artifacts are published.  This is currently how OpenShift CLIs are published.  The ConsoleCLIDownloads
+CRs reference this index.  The challenge is offering these in disconnected environments and automating the download and extraction of the artifacts.  
+
+3.  Images for each CLI. The CLI in-cluster manager would provide an extract mechanism for each CLI image, similar to how we currently 
+`oc adm release extract --command oc` and `oc image extract`.   A Custom Resource for each CLI would provide the in-cluster CLI manager information about each CLI and its image.
+In disconnected, the images will be available through a mirrored local registry.  The logic for `oc image extract` can be re-used and extended for extracting the CLIs. 
+    * CLI owners build images to package artifacts similar to https://github.com/sallyom/clustercli/tree/master/hack/cli-images
```

Comment:
The examples does not seem to package sha256 signatures for the binaries to images. This should consider the man-in-the-middle attacks and make them available.

Reply:
yes, the sha256sum will be included - same as it's included with https://mirror.openshift.com/pub/openshift-v4/clients